core/wrapper/socestimator.go

```diff
@@ -0,0 +1,92 @@
+package wrapper
+
+import (
+	"math"
+	"time"
+
+	"github.com/andig/evcc/api"
+	"github.com/andig/evcc/util"
+)
+
+// SocEstimator provides vehicle soc and charge duration
+// Vehicle SoC can be estimated to provide more granularity
+type SocEstimator struct {
+	log      *util.Logger
+	vehicle  api.Vehicle
+	estimate bool
+
+	capacity          float64 // vehicle capacity in Wh cached to simplify testing
+	socCharge         float64 // estimated vehicle SoC
+	prevSoC           float64 // previous vehicle SoC in %
+	prevChargedEnergy float64 // previous charged energy in Wh
+	energyPerSocStep  float64 // Energy per SoC percent in Wh
+}
+
+// NewSocEstimator creates new estimator
+func NewSocEstimator(log *util.Logger, vehicle api.Vehicle, estimate bool) *SocEstimator {
+	s := &SocEstimator{
+		log:      log,
+		vehicle:  vehicle,
+		estimate: estimate,
+	}
+
+	s.Reset()
+
+	return s
+}
+
+// Reset resets the estimation process to default values
+func (s *SocEstimator) Reset() {
+	s.prevSoC = -1
+	s.prevChargedEnergy = 0
+	s.capacity = float64(s.vehicle.Capacity()) * 1e3 // cache to simplify debugging
+	s.energyPerSocStep = s.capacity / 100
+}
+
+// RemainingChargeDuration returns the remaining duration estimate based on SoC, target and charge power
+func (s *SocEstimator) RemainingChargeDuration(chargePower float64, targetSoC int) time.Duration {
+	if chargePower > 0 {
+		percentRemaining := float64(targetSoC) - s.socCharge
+		if percentRemaining <= 0 {
+			return 0
+		}
+
+		whRemaining := percentRemaining / 100 * s.capacity
+		return time.Duration(float64(time.Hour) * whRemaining / chargePower).Round(time.Second)
+	}
+
+	return -1
+}
+
+// SoC implements Vehicle.ChargeState with addition of given charged energy
+func (s *SocEstimator) SoC(chargedEnergy float64) (float64, error) {
+	f, err := s.vehicle.ChargeState()
+	if err != nil {
+		return s.socCharge, err
+	}
+
+	s.socCharge = f
+
+	if s.estimate {
+		socDelta := f - s.prevSoC
```

Comment:
Hier wird prevSoC abgezogen, beim ersten Durchlauf also -1. Warum wird in diesem Fall -1 abgezogen? Im ersten Durchlauf gibt es kein Delta und der Gradient kann noch gar nicht berechnet werden?

Reply:
Ist behoben. prevSoc wird nun bei Reset auf 0 initialisiert. prevSoc muss aber ohnehin > 0 sein.